lib/producer/higher-level-producer.js

```diff
@@ -0,0 +1,201 @@
+    // mode we don't need to do multiple calls to poll since they all will yield
+    // the same result
+    pollingRefTimeout: null,
+  };
+
+  // Add the polling ref counter to the class which ensures we poll when we go active
+  this._hl.pollingRef = new RefCounter(function() {
+    self._hl.pollingRefTimeout = setInterval(function() {
+      try {
+        self.poll();
+      } catch (e) {
+        if (!self._isConnected) {
+          // If we got disconnected for some reason there is no point
+          // in polling anymore
+          clearInterval(self._hl.pollingRefTimeout);
+        }
+      }
+    }, 1);
+  }, function() {
+    clearInterval(self._hl.pollingRefTimeout);
+  });
+
+  // Default poll interval. More sophisticated polling is also done in create rule method
+  this.setPollInterval(1000);
+
+  // Listen to all delivery reports to propagate elements with a _message_id to the emitter
+  this.on('delivery-report', function(err, report) {
+    if (report.opaque && report.opaque.__message_id !== undefined) {
+      self._hl.deliveryEmitter.emit(report.opaque.__message_id, err, report.offset);
+    }
+  });
+
+  // Save old produce here since we are making some modifications for it
+  this._oldProduce = this.produce;
+  this.produce = this._modifiedProduce;
+
+  // Serializer information
+  this.keySerializer = noopSerializer;
+  this.valueSerializer = noopSerializer;
+}
+
+/**
+ * Produce a message to Kafka asynchronously.
+ *
+ * This is the method mainly used in this class. Use it to produce
+ * a message to Kafka.
+ *
+ * When this is sent off, and you recieve your callback, the assurances afforded
+ * to you will be equal to those provided by your ack level.
+ *
+ * @param {string} topic - The topic name to produce to.
+ * @param {number|null} partition - The partition number to produce to.
+ * @param {Buffer|null} message - The message to produce.
+ * @param {string} key - The key associated with the message.
+ * @param {number|null} timestamp - Timestamp to send with the message.
+ * @param {function} callback - Callback to call when the delivery report is recieved.
+ * @throws {LibrdKafkaError} - Throws a librdkafka error if it failed.
+ * @return {boolean} - returns an error if it failed, or true if not
+ * @see Producer#produce
+ */
+HighLevelProducer.prototype._modifiedProduce = function(topic, partition, message, key, timestamp, callback) {
+  // Add the message id
+  var opaque = {
+    __message_id: this._hl.messageId++,
+  };
+
+  this._hl.pollingRef.increment();
+
+  var self = this;
+
+  this._hl.deliveryEmitter.once(opaque.__message_id, function(err, offset) {
+    self._hl.pollingRef.decrement();
+    setImmediate(function() {
+      // Offset must be greater than or equal to 0 otherwise it is a null offset
+      // Possibly because we have acks off
+      callback(err, offset >= 0 ? offset : null);
+    });
+  });
+
+  return this._oldProduce(topic, partition,
```

Comment:
If we did move to some kind of async `*Serializer` then this logic would have to change to get the serialized values and only produce once the serialization is complete (or blows up)